### PR TITLE
Add retry logic to fakerp

### LINKED ISSUE #1122: retry code function - fakeRP utils
MSFT are retrying code in multiple places, based on error codes. We need something similar in our fakeRP.  This should be aligned with MSFT code, so they could reuse our error codes, or return results and wrap our plugin code functions with their side "retry functions"/ 

MSFT comment:
``` 
generally, we have a retry function that takes two functions. One outputs err, the other accepts the error and decides if it an error that should be retried
```
If we built something under these lines MSFT could reuse this on their end too. 

We should retry anything that goes "over the internet" in our code. Like kube api calls, arm calls, etc.

### PR DESCRIPTION
**Pull request by charlesakalugwu**
This work adds retry logic which works depending on the type of error emitted by the function to be retried. 

- [x] Add retry logic and tests
- [x] Refactor fakerp and clients to use retry logic
- [x] Add other error codes from #1122 
```release-note
NONE
```

closes #1122

### CODE REVIEW

File: hack/setup-rbac/main.go [hunk 1]
```diff
@@ -2,71 +2,25 @@ package main
...
 	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
 
 	"github.com/openshift/openshift-azure/pkg/fakerp"
 	"github.com/openshift/openshift-azure/pkg/util/azureclient"
 )
 
-// listActions lists all the mutating actions carried out on a resource group in
-// the last 6 hours, along with the principal that carried out the action
-func listActions(ctx context.Context, resourceGroupName string) error {
```
Comment (mjudeikis) [inline hack/setup-rbac/main.go]: @jim-minter was this intentional removal?

**Comment by mjudeikis** [comment]
/lgtm
Any other errors code can be taken as separate PR. this should help to drive flakes down so merging.

File: pkg/util/azureclient/resources.go [hunk 2]
```diff
@@ -67,9 +68,9 @@ type resourcesClient struct {
 var _ ResourcesClient = &resourcesClient{}
 
 // NewResourcesClient creates a new ResourcesClient
-func NewResourcesClient(ctx context.Context, subscriptionID string, authorizer autorest.Authorizer) ResourcesClient {
+func NewResourcesClient(ctx context.Context, log *logrus.Entry, subscriptionID string, authorizer autorest.Authorizer) ResourcesClient {
 	client := resources.NewClient(subscriptionID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "Client", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/resources.go]:
```suggestion
	setupClient(ctx, log, "ResourcesClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/network.go [hunk 3]
```diff
@@ -50,9 +51,9 @@ type virtualNetworkPeeringsClient struct {
 var _ VirtualNetworksPeeringsClient = &virtualNetworkPeeringsClient{}
 
 // NewVirtualNetworksPeeringsClient creates a new VirtualMachineScaleSetVMsClient
-func NewVirtualNetworksPeeringsClient(ctx context.Context, subscriptionID string, authorizer autorest.Authorizer) VirtualNetworksPeeringsClient {
+func NewVirtualNetworksPeeringsClient(ctx context.Context, log *logrus.Entry, subscriptionID string, authorizer autorest.Authorizer) VirtualNetworksPeeringsClient {
 	client := network.NewVirtualNetworkPeeringsClient(subscriptionID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "VirtualNetworkPeeringsClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/network.go]:
```suggestion
	setupClient(ctx, log, "VirtualNetworksPeeringsClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/network.go [hunk 4]
```diff
@@ -23,9 +24,9 @@ type virtualNetworksClient struct {
 var _ VirtualNetworksClient = &virtualNetworksClient{}
 
 // NewVirtualNetworkClient creates a new VirtualNetworkClient
-func NewVirtualNetworkClient(ctx context.Context, subscriptionID string, authorizer autorest.Authorizer) VirtualNetworksClient {
+func NewVirtualNetworkClient(ctx context.Context, log *logrus.Entry, subscriptionID string, authorizer autorest.Authorizer) VirtualNetworksClient {
 	client := network.NewVirtualNetworksClient(subscriptionID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "VirtualNetworksClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/network.go]:
```suggestion
	setupClient(ctx, log, "VirtualNetworkClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/keyvault.go [hunk 5]
```diff
@@ -19,9 +20,9 @@ type vaultsMgmtClient struct {
 var _ VaultMgmtClient = &vaultsMgmtClient{}
 
 // NewVaultMgmtClient get a new client for management actions
-func NewVaultMgmtClient(ctx context.Context, subscriptionID string, authorizer autorest.Authorizer) VaultMgmtClient {
+func NewVaultMgmtClient(ctx context.Context, log *logrus.Entry, subscriptionID string, authorizer autorest.Authorizer) VaultMgmtClient {
 	client := mgmtkeyvault.NewVaultsClient(subscriptionID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "VaultsClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/keyvault.go]:
```suggestion
	setupClient(ctx, log, "VaultMgmtClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/keyvault.go [hunk 6]
```diff
@@ -42,9 +43,9 @@ var _ KeyVaultClient = &keyVaultsClient{}
 // NewKeyVaultClient gets a new client for accessing vault values.  Important:
 // the authorizer supplied must have its resource set to
 // "https://vault.azure.net"
-func NewKeyVaultClient(ctx context.Context, authorizer autorest.Authorizer) KeyVaultClient {
+func NewKeyVaultClient(ctx context.Context, log *logrus.Entry, authorizer autorest.Authorizer) KeyVaultClient {
 	client := keyvault.New()
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "BaseClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/keyvault.go]:
```suggestion
	setupClient(ctx, log, "KeyVaultClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/graphrbac.go [hunk 7]
```diff
@@ -22,9 +23,9 @@ type rbacApplicationsClient struct {
 var _ RBACApplicationsClient = &rbacApplicationsClient{}
 
 // NewRBACApplicationsClient creates a new ApplicationsClient
-func NewRBACApplicationsClient(ctx context.Context, tenantID string, authorizer autorest.Authorizer) RBACApplicationsClient {
+func NewRBACApplicationsClient(ctx context.Context, log *logrus.Entry, tenantID string, authorizer autorest.Authorizer) RBACApplicationsClient {
 	client := graphrbac.NewApplicationsClient(tenantID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "ApplicationsClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/graphrbac.go]:
```suggestion
	setupClient(ctx, log, "RBACApplicationsClient", &client.Client, authorizer)
```

File: pkg/util/azureclient/graphrbac.go [hunk 8]
```diff
@@ -43,9 +44,9 @@ type rbacGroupsClient struct {
 var _ RBACGroupsClient = &rbacGroupsClient{}
 
 // NewRBACApplicationsClient creates a new ApplicationsClient
-func NewRBACGroupsClient(ctx context.Context, tenantID string, authorizer autorest.Authorizer) RBACGroupsClient {
+func NewRBACGroupsClient(ctx context.Context, log *logrus.Entry, tenantID string, authorizer autorest.Authorizer) RBACGroupsClient {
 	client := graphrbac.NewGroupsClient(tenantID)
-	setupClient(ctx, &client.Client, authorizer)
+	setupClient(ctx, log, "GroupsClient", &client.Client, authorizer)
```
Comment (charlesakalugwu) [inline pkg/util/azureclient/graphrbac.go]:
```suggestion
	setupClient(ctx, log, "RBACGroupsClient", &client.Client, authorizer)
```

File: hack/setup-rbac/main.go (hunk 1 above)
Comment (charlesakalugwu) [reply hack/setup-rbac/main.go]: @mjudeikis seems this was not being used anywhere

Comment (jim-minter) [reply hack/setup-rbac/main.go]: yeah, it was annoying me since the insights change.  I don't think we're going to use that code again.

**Comment by jim-minter** [comment]
rebased
@charlesakalugwu I've made the names suck less

**Pull request merged**